File: Python/maze_utils/generator.py
"""
Maze Generators - 迷宫生成算法
==============================

实现多种经典的迷宫生成算法：
- DFS（深度优先搜索）
- Prim算法
- Kruskal算法
- 递归分割
- Eller算法
- 二叉树算法
"""

import random
from typing import Optional, List, Tuple, Set
from .maze import Maze, Cell, Direction


def generate_dfs(width: int, height: int, seed: Optional[int] = None) -> Maze:
    """
    使用深度优先搜索（递归回溯）生成迷宫
    
    特点：
    - 生成长廊较多
    - 偏差较大，路径曲折
    - 实现简单，效率高
    
    Args:
        width: 迷宫宽度
        height: 迷宫高度
        seed: 随机种子（用于复现）
        
    Returns:
        生成的迷宫
    """
    if seed is not None:
        random.seed(seed)
    
    maze = Maze(width, height)
    
    # 使用栈实现非递归DFS
    stack = [(0, 0)]
    maze.get_cell(0, 0).visited = True
    
    while stack:
        x, y = stack[-1]
        
        # 获取未访问的邻居
        neighbors = maze.get_unvisited_neighbors(x, y)
        
        if neighbors:
            # 随机选择一个邻居
            direction, (nx, ny) = random.choice(neighbors)
            
            # 移除墙壁
            maze.remove_wall(x, y, nx, ny)
            
            # 标记访问并压栈
            maze.get_cell(nx, ny).visited = True
            stack.append((nx, ny))
        else:
            # 回溯
            stack.pop()
    
    maze.reset_visited()
    return maze


def generate_prim(width: int, height: int, seed: Optional[int] = None) -> Maze:
    """
    使用Prim算法生成迷宫
    
    特点：
    - 生成更均匀的迷宫
    - 路径分支多，较短的死胡同
    - 类似于最小生成树
    
    Args:
        width: 迷宫宽度
        height: 迷宫高度
        seed: 随机种子
        
    Returns:
        生成的迷宫
    """
    if seed is not None:
        random.seed(seed)
    
    maze = Maze(width, height)
    
    # 从起点开始
    start_x, start_y = 0, 0
    maze.get_cell(start_x, start_y).visited = True
    
    # 候选边列表：(当前单元格, 邻居单元格)
    frontier = []
    
    def add_frontier(x: int, y: int) -> None:
        for _, (nx, ny) in maze.get_unvisited_neighbors(x, y):
            frontier.append(((x, y), (nx, ny)))
    
    add_frontier(start_x, start_y)
    
    while frontier:
        # 随机选择一条边
        idx = random.randint(0, len(frontier) - 1)
        (x, y), (nx, ny) = frontier.pop(idx)
        
        if maze.get_cell(nx, ny).visited:
            continue
        
        # 连接单元格
        maze.remove_wall(x, y, nx, ny)
        maze.get_cell(nx, ny).visited = True
        add_frontier(nx, ny)
    
    maze.reset_visited()
    return maze


def generate_kruskal(width: int, height: int, seed: Optional[int] = None) -> Maze:
    """
    使用Kruskal算法生成迷宫
    
    特点：
    - 生成均匀分布的迷宫
    - 没有明显的偏差
    - 基于最小生成树
    
    Args:
        width: 迷宫宽度
        height: 迷宫高度
        seed: 随机种子
        
    Returns:
        生成的迷宫
    """
    if seed is not None:
        random.seed(seed)
    
    maze = Maze(width, height)
    
    # 并查集
    parent = {(x, y): (x, y) for y in range(height) for x in range(width)}
    
    def find(pos: Tuple[int, int]) -> Tuple[int, int]:
        if parent[pos] != pos:
            parent[pos] = find(parent[pos])
        return parent[pos]
    
    def union(pos1: Tuple[int, int], pos2: Tuple[int, int]) -> bool:
        root1, root2 = find(pos1), find(pos2)
        if root1 != root2:
            parent[root1] = root2
            return True
        return False
    
    # 所有边
    edges = []
    for y in range(height):
        for x in range(width):
            for direction, (nx, ny) in maze.get_neighbors(x, y):
                if direction in (Direction.EAST, Direction.SOUTH):
                    edges.append(((x, y), (nx, ny)))
    
    random.shuffle(edges)
    
    for (x, y), (nx, ny) in edges:
        if union((x, y), (nx, ny)):
            maze.remove_wall(x, y, nx, ny)
    
    return maze


def generate_recursive_division(width: int, height: int, 
                                seed: Optional[int] = None) -> Maze:
    """
    使用递归分割算法生成迷宫
    
    特点：
    - 先创建空房间，再添加墙壁
        - 生成长直走廊
    - 视觉上有规律性
    
    Args:
        width: 迷宫宽度
        height: 迷宫高度
        seed: 随机种子
        
    Returns:
        生成的迷宫
    """
    if seed is not None:
        random.seed(seed)
    
    maze = Maze(width, height)
    
    # 先移除所有内部墙壁
    for y in range(height):
        for x in range(width):
            cell = maze.get_cell(x, y)
            cell.walls = set()
    
    def divide(x: int, y: int, w: int, h: int) -> None:
        if w < 2 or h < 2:
            return
        
        # 选择分割方向
        if w > h:
            # 垂直分割
            wall_x = x + random.randint(1, w - 1)
            passage_y = random.randint(y, y + h - 1)
            
            for py in range(y, y + h):
                if py != passage_y and wall_x < width:
                    maze.add_wall(wall_x - 1, py, wall_x, py)
            
            divide(x, y, wall_x - x, h)
            divide(wall_x, y, x + w - wall_x, h)
        else:
            # 水平分割
            wall_y = y + random.randint(1, h - 1)
            passage_x = random.randint(x, x + w - 1)
            
            for px in range(x, x + w):
                if px != passage_x and wall_y < height:
                    maze.add_wall(px, wall_y - 1, px, wall_y)
            
            divide(x, y, w, wall_y - y)
            divide(x, wall_y, w, y + h - wall_y)
    
    divide(0, 0, width, height)
    return maze


def generate_ellers(width: int, height: int, seed: Optional[int] = None) -> Maze:
    """
    使用Eller算法生成迷宫
    
    特点：
    - 逐行生成，内存效率高
    - 可以生成无限大的迷宫
    - 产生适中的纹理
    
    Args:
        width: 迷宫宽度
        height: 迷宫高度
        seed: 随机种子
        
    Returns:
        生成的迷宫
    """
    if seed is not None:
        random.seed(seed)
    
    maze = Maze(width, height)
    
    # 每行的集合标记
    sets = [i for i in range(width)]
    next_set = width
    
    for y in range(height):
        # 水平连接：随机合并相邻不同集合的单元格
        for x in range(width - 1):
            if sets[x] != sets[x + 1] and (y == height - 1 or random.random() < 0.5):
                # 合并集合
                old_set = sets[x + 1]
                new_set = sets[x]
                for i in range(width):
                    if sets[i] == old_set:
                        sets[i] = new_set
                # 移除墙壁
                maze.remove_wall(x, y, x + 1, y)
        
        if y == height - 1:
            continue
        
        # 垂直连接：每个集合至少向下连接一次
        verticals = {}
        for x in range(width):
            s = sets[x]
            if s not in verticals:
                verticals[s] = []
            verticals[s].append(x)
        
        new_sets = [-1] * width
        
        for s, columns in verticals.items():
            # 随机选择要向下连接的单元格
            num_down = random.randint(1, len(columns))
            down_cols = random.sample(columns, num_down)
            
            for col in down_cols:
                maze.remove_wall(col, y, col, y + 1)
                new_sets[col] = s
        
        # 分配新集合给未连接的单元格
        for x in range(width):
            if new_sets[x] == -1:
                new_sets[x] = next_set
                next_set += 1
        
        sets = new_sets
    
    return maze


def generate_binary_tree(width: int, height: int, 
                         seed: Optional[int] = None,
                         bias: str = 'northeast') -> Maze:
    """
    使用二叉树算法生成迷宫
    
    特点：
    - 最简单的迷宫算法之一
    - 有明显的对角线偏差
    - 非常快速
    
    Args:
        width: 迷宫宽度
        height: 迷宫高度
        seed: 随机种子
        bias: 偏差方向 ('northeast', 'northwest', 'southeast', 'southwest')
        
    Returns:
        生成的迷宫
    """
    if seed is not None:
        random.seed(seed)
    
    maze = Maze(width, height)
    
    # 根据偏差方向确定主方向和次方向
    biases = {
        'northeast': (Direction.NORTH, Direction.EAST),
        'northwest': (Direction.NORTH, Direction.WEST),
        'southeast': (Direction.SOUTH, Direction.EAST),
        'southwest': (Direction.SOUTH, Direction.WEST),
    }
    
    primary, secondary = biases.get(bias, biases['northeast'])
    
    for y in range(height):
        for x in range(width):
            # 获取可用的方向
            options = []
            
            if primary == Direction.NORTH and y > 0:
                options.append((x, y, x, y - 1))
            elif primary == Direction.SOUTH and y < height - 1:
                options.append((x, y, x, y + 1))
            elif primary == Direction.EAST and x < width - 1:
                options.append((x, y, x + 1, y))
            elif primary == Direction.WEST and x > 0:
                options.append((x, y, x - 1, y))
            
            if secondary == Direction.NORTH and y > 0:
                options.append((x, y, x, y - 1))
            elif secondary == Direction.SOUTH and y < height - 1:
                options.append((x, y, x, y + 1))
            elif secondary == Direction.EAST and x < width - 1:
                options.append((x, y, x + 1, y))
            elif secondary == Direction.WEST and x > 0:
                options.append((x, y, x - 1, y))
            
            if options:
                # 随机选择一个方向
                x1, y1, x2, y2 = random.choice(options)
                maze.remove_wall(x1, y1, x2, y2)
    
    return maze


def generate_sidewinder(width: int, height: int, seed: Optional[int] = None) -> Maze:
    """
    使用Sidewinder算法生成迷宫
    
    特点：
    - 简单高效
    - 有水平偏差
    - 顶部行完全水平
    
    Args:
        width: 迷宫宽度
        height: 迷宫高度
        seed: 随机种子
        
    Returns:
        生成的迷宫
    """
    if seed is not None:
        random.seed(seed)
    
    maze = Maze(width, height)
    
    # 顶行：全部向东连通
    for x in range(width - 1):
        maze.remove_wall(x, 0, x + 1, 0)
    
    # 其余行
    for y in range(1, height):
        run_start = 0
        
        for x in range(width):
            # 向东延伸或向北
            if x == width - 1 or random.random() < 0.5:
                # 从当前run中随机选择一个单元格向北连接
                north_x = random.randint(run_start, x)
                maze.remove_wall(north_x, y, north_x, y - 1)
                run_start = x + 1
            else:
                # 向东延伸
                maze.remove_wall(x, y, x + 1, y)
    
    return maze

File: Go/maze_utils/generator.go
package maze_utils

import (
	"math/rand"
	"sort"
	"time"
)

// 生成器配置
type GeneratorConfig struct {
	Seed int64
}

// initRandom 初始化随机数生成器
func initRandom(seed int64) *rand.Rand {
	if seed == 0 {
		seed = time.Now().UnixNano()
	}
	return rand.New(rand.NewSource(seed))
}

// GenerateDFS 使用深度优先搜索算法生成迷宫
// 特点：生成长廊，路径曲折
func GenerateDFS(width, height int, seed int64) *Maze {
	m := NewMaze(width, height)
	m.Seed = seed
	r := initRandom(seed)

	visited := make(map[[2]int]bool)
	var stack [][2]int

	// 从起点开始
	start := [2]int{0, 0}
	stack = append(stack, start)
	visited[start] = true

	for len(stack) > 0 {
		current := stack[len(stack)-1]
		x, y := current[0], current[1]

		// 获取未访问的邻居
		neighbors := getUnvisitedNeighbors(m, x, y, visited)
		if len(neighbors) == 0 {
			stack = stack[:len(stack)-1]
			continue
		}

		// 随机选择一个邻居
		next := neighbors[r.Intn(len(neighbors))]
		nx, ny := next[0], next[1]

		// 移除墙
		m.RemoveWall(x, y, nx, ny)

		// 标记为已访问
		visited[next] = true
		stack = append(stack, next)
	}

	return m
}

// GeneratePrim 使用 Prim 算法生成迷宫
// 特点：均匀分布，分支多
func GeneratePrim(width, height int, seed int64) *Maze {
	m := NewMaze(width, height)
	m.Seed = seed
	r := initRandom(seed)

	inMaze := make(map[[2]int]bool)
	walls := make([][4]int, 0) // [x1, y1, x2, y2]

	// 从起点开始
	start := [2]int{0, 0}
	inMaze[start] = true

	// 添加起点的所有墙
	walls = addWalls(m, 0, 0, walls)

	for len(walls) > 0 {
		// 随机选择一面墙
		idx := r.Intn(len(walls))
		wall := walls[idx]
		walls = append(walls[:idx], walls[idx+1:]...)

		x1, y1, x2, y2 := wall[0], wall[1], wall[2], wall[3]
		cell1 := [2]int{x1, y1}
		cell2 := [2]int{x2, y2}

		// 如果只有一个单元格在迷宫中
		if inMaze[cell1] != inMaze[cell2] {
			// 移除墙
			m.RemoveWall(x1, y1, x2, y2)

			// 将新单元格加入迷宫
			if !inMaze[cell1] {
				inMaze[cell1] = true
				walls = addWalls(m, x1, y1, walls)
			}
			if !inMaze[cell2] {
				inMaze[cell2] = true
				walls = addWalls(m, x2, y2, walls)
			}
		}
	}

	return m
}

// GenerateKruskal 使用 Kruskal 算法生成迷宫
// 特点：无偏差的均匀迷宫
func GenerateKruskal(width, height int, seed int64) *Maze {
	m := NewMaze(width, height)
	m.Seed = seed
	r := initRandom(seed)

	// 创建并查集
	parent := make(map[[2]int][2]int)
	rank := make(map[[2]int]int)

	// 初始化每个单元格
	cells := make([][2]int, 0)
	for y := 0; y < height; y++ {
		for x := 0; x < width; x++ {
			cell := [2]int{x, y}
			parent[cell] = cell
			rank[cell] = 0
			cells = append(cells, cell)
		}
	}

	// 查找函数 - 使用闭包前的变量声明以支持递归
	var find func([2]int) [2]int
	find = func(cell [2]int) [2]int {
		if parent[cell] != cell {
			parent[cell] = find(parent[cell])
		}
		return parent[cell]
	}

	// 合并函数
	union := func(cell1, cell2 [2]int) bool {
		root1 := find(cell1)
		root2 := find(cell2)
		if root1 == root2 {
			return false
		}
		if rank[root1] < rank[root2] {
			parent[root1] = root2
		} else if rank[root1] > rank[root2] {
			parent[root2] = root1
		} else {
			parent[root2] = root1
			rank[root1]++
		}
		return true
	}

	// 创建所有可能的墙（边）
	type edge struct {
		x1, y1, x2, y2 int
	}
	edges := make([]edge, 0)
	for y := 0; y < height; y++ {
		for x := 0; x < width; x++ {
			if x < width-1 {
				edges = append(edges, edge{x, y, x + 1, y})
			}
			if y < height-1 {
				edges = append(edges, edge{x, y, x, y + 1})
			}
		}
	}

	// 随机打乱边
	r.Shuffle(len(edges), func(i, j int) {
		edges[i], edges[j] = edges[j], edges[i]
	})

	// 处理每条边
	for _, e := range edges {
		cell1 := [2]int{e.x1, e.y1}
		cell2 := [2]int{e.x2, e.y2}
		if union(cell1, cell2) {
			m.RemoveWall(e.x1, e.y1, e.x2, e.y2)
		}
	}

	return m
}

// GenerateRecursiveDivision 使用递归分割算法生成迷宫
// 特点：先建房间再分割，长直走廊
func GenerateRecursiveDivision(width, height int, seed int64) *Maze {
	m := NewMaze(width, height)
	m.Seed = seed
	r := initRandom(seed)

	// 首先移除所有内部墙
	for y := 0; y < height; y++ {
		for x := 0; x < width; x++ {
			if x < width-1 {
				m.RemoveWall(x, y, x+1, y)
			}
			if y < height-1 {
				m.RemoveWall(x, y, x, y+1)
			}
		}
	}

	// 递归分割
	divide(m, r, 0, 0, width, height, chooseOrientation(width, height, r))

	return m
}

// chooseOrientation 选择分割方向
func chooseOrientation(width, height int, r *rand.Rand) bool {
	// true = 水平分割, false = 垂直分割
	if width < height {
		return true
	} else if height < width {
		return false
	}
	return r.Intn(2) == 0
}

// divide 递归分割
func divide(m *Maze, r *rand.Rand, x, y, width, height int, horizontal bool) {
	if width < 2 || height < 2 {
		return
	}

	// 选择分割位置（在区域中间选择一条线）
	var wx, wy int
	if horizontal {
		// 水平分割：分割线在 y 到 y+height-1 之间
		wy = y + r.Intn(height-1) + 1
		wx = x
	} else {
		// 垂直分割：分割线在 x 到 x+width-1 之间
		wx = x + r.Intn(width-1) + 1
		wy = y
	}

	// 选择通道位置（在分割线上选择一个位置不建墙）
	var px, py int
	if horizontal {
		// 水平分割线上的通道
		px = x + r.Intn(width)
		py = wy
	} else {
		// 垂直分割线上的通道
		px = wx
		py = y + r.Intn(height)
	}

	// 在分割线上建墙，除了通道位置
	if horizontal {
		// 水平分割：在分割线上方添加水平墙，分隔上下
		for i := 0; i < width; i++ {
			cx := x + i
			cy := wy
			if cx == px && cy == py {
				continue // 通道位置不建墙
			}
			// 添加水平墙：分隔 cy-1 和 cy
			if cy > y {
				m.AddWall(cx, cy-1, cx, cy)
			}
		}
	} else {
		// 垂直分割：在分割线左侧添加垂直墙，分隔左右
		for i := 0; i < height; i++ {
			cx := wx
			cy := y + i
			if cx == px && cy == py {
				continue // 通道位置不建墙
			}
			// 添加垂直墙：分隔 cx-1 和 cx
			if cx > x {
				m.AddWall(cx-1, cy, cx, cy)
			}
		}
	}

	// 递归分割子区域
	if horizontal {
		// 上半部分：y 到 wy-1
		divide(m, r, x, y, width, wy-y, chooseOrientation(width, wy-y, r))
		// 下半部分：wy 到 y+height-1
		divide(m, r, x, wy, width, y+height-wy, chooseOrientation(width, y+height-wy, r))
	} else {
		// 左半部分：x 到 wx-1
		divide(m, r, x, y, wx-x, height, chooseOrientation(wx-x, height, r))
		// 右半部分：wx 到 x+width-1
		divide(m, r, wx, y, x+width-wx, height, chooseOrientation(x+width-wx, height, r))
	}
}

// GenerateEllers 使用 Eller 算法生成迷宫
// 特点：逐行生成，内存高效，适合大迷宫
func GenerateEllers(width, height int, seed int64) *Maze {
	m := NewMaze(width, height)
	m.Seed = seed
	r := initRandom(seed)

	// 每行的集合编号
	sets := make([]int, width)
	// 集合计数器
	setCounter := 1

	// 初始化第一行
	for x := 0; x < width; x++ {
		sets[x] = setCounter
		setCounter++
	}

	// 逐行处理
	for y := 0; y < height; y++ {
		// 水平连接：随机合并同一行中相邻的不同集合
		for x := 0; x < width-1; x++ {
			if sets[x] != sets[x+1] {
				// 最后一行必须全部连接
				if y == height-1 || r.Intn(2) == 0 {
					m.RemoveWall(x, y, x+1, y)
					// 合并集合
					oldSet := sets[x+1]
					for i := 0; i < width; i++ {
						if sets[i] == oldSet {
							sets[i] = sets[x]
						}
					}
				}
			}
		}

		// 最后一行不需要垂直连接
		if y == height-1 {
			break
		}

		// 垂直连接：每个集合至少连接一次
		setMembers := make(map[int][]int)
		for x := 0; x < width; x++ {
			setMembers[sets[x]] = append(setMembers[sets[x]], x)
		}

		// 为每个集合创建垂直连接
		nextSets := make([]int, width)
		for i := range nextSets {
			nextSets[i] = 0
		}

		for set, members := range setMembers {
			// 随机选择至少一个成员向下连接
			sort.Ints(members)
			numDown := 1 + r.Intn(len(members))
			perm := r.Perm(len(members))
			for i := 0; i < numDown; i++ {
				x := members[perm[i]]
				m.RemoveWall(x, y, x, y+1)
				nextSets[x] = set
			}
		}

		// 为未连接的单元格分配新集合
		for x := 0; x < width; x++ {
			if nextSets[x] == 0 {
				nextSets[x] = setCounter
				setCounter++
			}
		}

		sets = nextSets
	}

	return m
}

// GenerateBinaryTree 使用二叉树算法生成迷宫
// 特点：最简单的算法，有对角偏差
// bias: 'N'/'E' = 偏向北和东, 'N'/'W' = 偏向北和西, 'S'/'E' = 偏向南和东, 'S'/'W' = 偏向南和西
func GenerateBinaryTree(width, height int, seed int64, bias string) *Maze {
	m := NewMaze(width, height)
	m.Seed = seed
	r := initRandom(seed)

	// 默认偏北和东
	dir1, dir2 := North, East
	if len(bias) >= 2 {
		switch bias[0] {
		case 'N', 'n':
			dir1 = North
		case 'S', 's':
			dir1 = South
		}
		switch bias[1] {
		case 'E', 'e':
			dir2 = East
		case 'W', 'w':
			dir2 = West
		}
	}

	// 方向偏移
	offsets := map[Direction][2]int{
		North: {0, -1},
		South: {0, 1},
		East:  {1, 0},
		West:  {-1, 0},
	}

	for y := 0; y < height; y++ {
		for x := 0; x < width; x++ {
			// 随机选择方向
			choices := make([]Direction, 0)

			// 检查方向1是否可行
			dx1, dy1 := offsets[dir1][0], offsets[dir1][1]
			nx1, ny1 := x+dx1, y+dy1
			if nx1 >= 0 && nx1 < width && ny1 >= 0 && ny1 < height {
				choices = append(choices, dir1)
			}

			// 检查方向2是否可行
			dx2, dy2 := offsets[dir2][0], offsets[dir2][1]
			nx2, ny2 := x+dx2, y+dy2
			if nx2 >= 0 && nx2 < width && ny2 >= 0 && ny2 < height {
				choices = append(choices, dir2)
			}

			// 随机选择一个方向
			if len(choices) > 0 {
				dir := choices[r.Intn(len(choices))]
				var nx, ny int
				if dir == dir1 {
					nx, ny = nx1, ny1
				} else {
					nx, ny = nx2, ny2
				}
				m.RemoveWall(x, y, nx, ny)
			}
		}
	}

	return m
}

// getUnvisitedNeighbors 获取未访问的邻居
func getUnvisitedNeighbors(m *Maze, x, y int, visited map[[2]int]bool) [][2]int {
	var neighbors [][2]int
	directions := [][2]int{{0, -1}, {1, 0}, {0, 1}, {-1, 0}}

	for _, d := range directions {
		nx, ny := x+d[0], y+d[1]
		if nx >= 0 && nx < m.Width && ny >= 0 && ny < m.Height {
			if !visited[[2]int{nx, ny}] {
				neighbors = append(neighbors, [2]int{nx, ny})
			}
		}
	}
	return neighbors
}

// addWalls 添加单元格的所有墙到列表
func addWalls(m *Maze, x, y int, walls [][4]int) [][4]int {
	directions := [][2]int{{0, -1}, {1, 0}, {0, 1}, {-1, 0}}

	for _, d := range directions {
		nx, ny := x+d[0], y+d[1]
		if nx >= 0 && nx < m.Width && ny >= 0 && ny < m.Height {
			walls = append(walls, [4]int{x, y, nx, ny})
		}
	}
	return walls
}